File: 2551-apply-operations-to-an-array/2551-apply-operations-to-an-array.cpp
class Solution {
public:
    vector<int> applyOperations(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans;
        int count = 0;
        int i = 0;

        bool lastChangedToZero = false;
        while(i<n-1){
            if(nums[i]==0){
                count++;
                i++;
            }
            else if(nums[i]==nums[i+1]){
                if(i==n-2){
                    lastChangedToZero = true;
                }
                ans.push_back(nums[i]*2);
                count++;
                i+=2;
            }
            else{
                ans.push_back(nums[i]);
                i++;
            }
        }
        if(!lastChangedToZero){
            ans.push_back(nums[n-1]);
        }
        while(count--){
            ans.push_back(0);
        }
        return ans;
    }
};

File: 2551-apply-operations-to-an-array/2551-apply-operations-to-an-array.rs
impl Solution {  //declation of a implementation.
    pub fn apply_operations(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut ans = Vec::new();  //mut - mutable. Default it is immutable.
        let mut count = 0;
        let mut i = 0;
        let mut lastChangedToZero = false;

        while i<n-1 {
            if nums[i] == 0 {
                count+=1;
                i+=1;
            }
            else if nums[i] == nums[i+1] {
                if i == n-2 {
                    lastChangedToZero = true;
                }
                ans.push(nums[i]*2);
                count+=1;
                i+=2;
            }
            else{
                ans.push(nums[i]);
                i+=1;
            }
        }
        if !lastChangedToZero {
            ans.push(nums[n-1]);
        }

        for _ in 0..count {
            ans.push(0);
        }
        return ans;
    }
}